src/jit/lowerxarch.cpp

```diff
@@ -978,6 +978,63 @@ void Lowering::LowerHWIntrinsic(GenTreeHWIntrinsic* node)
             LowerHWIntrinsicCC(node, NI_AVX_PTEST, GenCondition::UGT);
             break;
 
+        case NI_FMA_MultiplyAddScalar:
+        {
+            // Math(F).FusedMultiplyAdd is expanded into NI_FMA_MultiplyAddScalar but
+            // depending on additional GT_NEG nodes it can be:
+            //
+            //    x *  y + z -> NI_FMA_MultiplyAddScalar
+            //    x * -y + z -> NI_FMA_MultiplyAddNegatedScalar
+            //   -x *  y + z -> NI_FMA_MultiplyAddNegatedScalar
+            //   -x * -y + z -> NI_FMA_MultiplyAddScalar
+            //    x *  y - z -> NI_FMA_MultiplySubtractScalar
+            //    x * -y - z -> NI_FMA_MultiplySubtractNegatedScalar
+            //   -x *  y - z -> NI_FMA_MultiplySubtractNegatedScalar
+            //   -x * -y - z -> NI_FMA_MultiplySubtractScalar
+            //
+            GenTreeArgList* argList = node->gtGetOp1()->AsArgList();
+            if (argList->Current()->OperIsHWIntrinsic()) // CreateScalarUnsafe is expected
+            {
+                GenTreeUnOp* createScalarOpX = argList->Current()->AsUnOp();
+                GenTree*     argX            = createScalarOpX->gtGetOp1();
+                argList                      = argList->Rest();
+                if (argList->Current()->OperIsHWIntrinsic()) // CreateScalarUnsafe is expected
```

Comment:
I don't think that this repetition and associated nesting is necessary. You should be able to extract the operands to a 3 element local array and work with that.

Reply:
Could you please take a look at my latest commit - do you still would prefer an array and by the way which data structure is usually used for such cases?

Reply:
Thanks, it would better to use an array. Using a loop and then writing different code inside the loop for every possible index is kind of strange (and definitely unnecessary in this case). Also note that the assembly mnemonics use indices too, though 1 based. You could make the array one element larger and use only elements 1-3 so there's no "which one is which" kind of confusion.